operators/constellation-node-operator/internal/executor/executor.go

```diff
@@ -0,0 +1,215 @@
+	if e.running {
+		return stopWait
+	}
+	e.running = true
+
+	e.externalTrigger = make(chan struct{}, 1)
+	e.stop = make(chan struct{}, 1)
+	// execute is used by the go routines below to communicate
+	// that a reconciliation should happen
+	execute := make(chan struct{}, 1)
+	// nextScheduledReconcile is used to communicate the next scheduled reconciliation time
+	nextScheduledReconcile := make(chan time.Duration, 1)
+	// trigger a reconciliation on startup
+	nextScheduledReconcile <- 0
+
+	wg.Add(2)
+
+	// timer routine is responsible for triggering the reconciliation after the timer expires
+	// or when triggered externally
+	go func() {
+		defer wg.Done()
+		defer close(execute)
+		defer logr.Info("Timer stopped")
+		for {
+			nextScheduledReconcileAfter := <-nextScheduledReconcile
+			timer := *time.NewTimer(nextScheduledReconcileAfter)
+			select {
+			case <-e.stop:
+				return
+			case <-ctx.Done():
+				return
+			case <-e.externalTrigger:
+			case <-timer.C:
+			}
+			execute <- struct{}{}
+		}
+	}()
+
+	// executor routine is responsible for executing the reconciliation
+	go func() {
+		defer func() {
+			e.mux.Lock()
+			defer e.mux.Unlock()
+			e.running = false
+		}()
+		defer wg.Done()
+		defer close(nextScheduledReconcile)
+		defer logr.Info("Executor stopped")
+
+		for {
+			_, ok := <-execute
+			// execute channel closed. executor should stop
+			if !ok {
+				return
+			}
+			res, err := e.controller.Reconcile(ctx)
+			var requeueAfter time.Duration
+			switch {
+			case err != nil:
+				logr.Error(err, "reconciliation failed")
+				requeueAfter = e.rateLimiter.When(rateLimiterItem) // requeue with rate limiter
+			case res.Requeue && res.RequeueAfter != 0:
+				e.rateLimiter.Forget(rateLimiterItem) // reset the rate limiter
+				requeueAfter = res.RequeueAfter       // requeue after the given duration
+			case res.Requeue:
+				requeueAfter = e.rateLimiter.When(rateLimiterItem) // requeue with rate limiter
+			default:
+				e.rateLimiter.Forget(rateLimiterItem) // reset the rate limiter
+				requeueAfter = e.pollingFrequency     // default polling frequency
+			}
+
+			nextScheduledReconcile <- requeueAfter
+		}
+	}()
+
+	return stopWait
+}
+
+// Stop stops the executor.
+// It does not block until the executor is stopped.
+func (e *Executor) Stop() {
+	select {
+	case e.stop <- struct{}{}:
+	default:
+	}
+	close(e.stop)
+}
+
+// Running returns true if the executor is running.
+// When the executor is stopped, it is not running anymore.
+func (e *Executor) Running() bool {
+	e.mux.Lock()
+	defer e.mux.Unlock()
+	return e.running
+}
+
+// Trigger triggers a reconciliation.
+// If a reconciliation is already pending, this call is a no-op.
+func (e *Executor) Trigger() {
+	select {
+	case e.externalTrigger <- struct{}{}:
+	default:
+	}
+}
+
+// Config is the configuration for the executor.
+type Config struct {
+	PollingFrequency time.Duration
+	RateLimiter      RateLimiter
+}
+
+func (c *Config) applyDefaults() {
```

Comment:
The reason I like this pattern is that the user can use a composite literal to initialize the struct which gives you named fields:

```go
conf := Config{
    PollingFrequency: 10 * time.Second,
    RateLimiter: customratelimiter.New(),
}
```

... and any uninitialized fields automatically get initialized by calling `applyDefaults()` as a consumer of the config.

`New` for config / options structs is a weird in my opinion since you lose the named fields:

```go
conf := New(10 * time.Second, customratelimiter.New())
```

And if you don't call an explicit constructor (`New` / `NewWithDefaults`) as a user, you may have default (nil) initialized interfaces that can lead to panics.

Reply:
I see your point. Normally, I'd suggest to simply use a config for `New()` to retain the naming, but we don't want to have configs all the way down, I suppose. 
I like your suggestion for the middle ground.